README.md
# proud-connect v0.0.3

> Connect server generating NPM downloads badges by author

[![NPM][proud-connect-icon]][proud-connect-url]

[![Build status][proud-connect-ci-image]][proud-connect-ci-url]
[![dependencies][proud-connect-dependencies-image]][proud-connect-dependencies-url]
[![devdependencies][proud-connect-devdependencies-image]][proud-connect-devdependencies-url]

[![endorse][endorse-image]][endorse-url]

[proud-connect-icon]: https://nodei.co/npm/proud-connect.png?downloads=true
[proud-connect-url]: https://npmjs.org/package/proud-connect
[proud-connect-ci-image]: https://travis-ci.org/bahmutov/proud-connect.png?branch=master
[proud-connect-ci-url]: https://travis-ci.org/bahmutov/proud-connect
[proud-connect-dependencies-image]: https://david-dm.org/bahmutov/proud-connect.png
[proud-connect-dependencies-url]: https://david-dm.org/bahmutov/proud-connect
[proud-connect-devdependencies-image]: https://david-dm.org/bahmutov/proud-connect/dev-status.png
[proud-connect-devdependencies-url]: https://david-dm.org/bahmutov/proud-connect#info=devDependencies
[endorse-image]: https://api.coderwall.com/bahmutov/endorsecount.png
[endorse-url]: https://coderwall.com/bahmutov



## Use

```
npm install
node index.js
```

Badge / json information can be fetched `http://localhost:3000/username

#### Using the deployed Heroku app

![bahmutov badge](http://proud.herokuapp.com/bahmutov)

Use Markdown:

    ![bahmutov badge](http://proud.herokuapp.com/bahmutov)

Or HTML:

    <img src="http://proud.herokuapp.com/bahmutov"></img>

to generate number of total NPM downloads in the last month.



### Related

* [proud](https://github.com/bahmutov/proud) is a stand alone CLI tool
* [proud-badge](https://github.com/bahmutov/proud-badge) generates badges
* [proud-heroku-app](https://github.com/bahmutov/proud-heroku-app) is
a *proud-connect* service [running on Heroku](http://proud.herokuapp.com/)

## Why?

Because one should be **proud** of his work.

### Small print

Author: Gleb Bahmutov &copy; 2013

* [@bahmutov](https://twitter.com/bahmutov)
* [glebbahmutov.com](http://glebbahmutov.com)
* [blog](http://bahmutov.calepin.co/)

License: MIT - do anything with the code, but don't blame me if it does not work.

Spread the word: tweet, star on github, click *endorse*, etc.

Support: if you find any problems with this module, email / tweet / open issue on Github



## History


0.0.3 / 2013-11-17
==================

  * you can pass desired format as /username/png or /username/json fixes #8
  * added unit tests
  * added travis file
  * added test as pre-commit
  * added simple sanity test

0.0.2 / 2013-11-17
==================

  * using grunt and grunt-readme, fixes #9, #10, #11
  * fixed concurrent problem

0.0.1 / 2013-11-13
==================

  * images older than day will be regenerated
  * returning badge and keeping it in memory
  * working badge return or text, depending on request




History.md

0.0.3 / 2013-11-17
==================

  * you can pass desired format as /username/png or /username/json fixes #8
  * added unit tests
  * added travis file
  * added test as pre-commit
  * added simple sanity test

0.0.2 / 2013-11-17
==================

  * using grunt and grunt-readme, fixes #9, #10, #11
  * fixed concurrent problem

0.0.1 / 2013-11-13
==================

  * images older than day will be regenerated
  * returning badge and keeping it in memory
  * working badge return or text, depending on request


src/test/app.js
var listen = require('../app');
var request = require('request');

var port = 4444;
var urlBase = 'http://localhost:' + port + '/';

gt.module('server', {
  setupOnce: function () {
    console.log('starting listening');
    listen(port);
  }
});

(function cachedPng() {
  var image;
  var took = 0;

  gt.async('/png', function () {
    var username = 'jashkenas';
    var opts = {
      url: urlBase + username + '/png'
    };
    var started = new Date();
    request(opts, function (err, response, body) {
      if (err) throw err;
      var finished = new Date();

      gt.equal(response.statusCode, 200, 'got response');
      gt.equal(response.headers['content-type'], 'image/png', 'returns png');
      image = body;
      gt.ok(image.length > 0, 'positive image length');
      took = finished - started;

      gt.ok(took > 0, 'took a few ms');
      gt.ok(response.headers['expires'], 'has expires header');
      gt.start();
    });
  });

  gt.async('/png 2 - cached', function () {
    var username = 'jashkenas';
    var opts = {
      url: urlBase + username + '/png'
    };
    var started = new Date();

    request(opts, function (err, response, body) {
      if (err) throw err;
      var finished = new Date();
      var took2 = finished - started;
      gt.ok(took2 < took, 'second request completed faster than first');

      gt.equal(response.statusCode, 200, 'got response');
      gt.start();
    });
  });
}());

gt.async('png has expires header', function () {
  var username = 'jashkenas';
  request(urlBase + username + '/png', function (err, response, body) {
    if (err) throw err;
    var finished = new Date();
    gt.equal(response.statusCode, 200, 'got response');
    gt.equal(response.headers['content-type'], 'image/png', 'returns png');
    gt.start();
  });
});

gt.async('no username', function () {
  var url = urlBase;
  request(url, function (err, response, body) {
    if (err) throw err;
    gt.equal(response.statusCode, 401, '401 without username');
    gt.start();
  });
});

gt.async('plain text', function () {
  var url = urlBase + 'jashkenas';
  request(url, function (err, response, body) {
    if (err) throw err;
    gt.equal(response.statusCode, 200, 'got response');
    gt.equal(response.headers['content-type'], 'text/plain', 'returns text');

    gt.ok(/jashkenas/.test(body), 'contains username');
    gt.ok(/Total/i.test(body), 'contains total');
    gt.start();
  });
});

gt.async('json', function () {
  var username = 'jashkenas';
  var opts = {
    url: urlBase + username,
    json: true
  };
  request(opts, function (err, response, body) {
    if (err) throw err;
    gt.equal(response.statusCode, 200, 'got response');
    gt.equal(response.headers['content-type'], 'application/json', 'returns json');

    gt.equal(body.username, username, 'correct username');
    gt.start();
  });
});

gt.async('/json', function () {
  var username = 'jashkenas';
  var opts = {
    url: urlBase + username + '/json'
  };
  request(opts, function (err, response, body) {
    if (err) throw err;
    gt.equal(response.statusCode, 200, 'got response');
    gt.equal(response.headers['content-type'], 'application/json', 'returns json');

    var result = JSON.parse(body);
    gt.equal(result.username, username, 'correct username');
    gt.start();
  });
});

// todo: make sure second request uses cached data
gt.async('/json 2', function () {
  var username = 'jashkenas';
  var opts = {
    url: urlBase + username + '/json'
  };
  request(opts, function (err, response, body) {
    if (err) throw err;
    gt.equal(response.statusCode, 200, 'got response');
    gt.equal(response.headers['content-type'], 'application/json', 'returns json');

    var result = JSON.parse(body);
    gt.equal(result.username, username, 'correct username');
    gt.start();
  });
});

gt.async('/json user without modules', function () {
  var username = 'lsilvo';
  var opts = {
    url: urlBase + username + '/json'
  };
  request(opts, function (err, response, body) {
    if (err) throw err;
    gt.equal(response.statusCode, 200, 'got response');
    gt.equal(response.headers['content-type'], 'application/json', 'returns json');

    var result = JSON.parse(body);
    gt.equal(result.username, username, 'correct username');
    gt.ok(!result.report, 'there is nothing to report');
    gt.start();
  });
});

gt.async('/png user without modules', function () {
  var username = 'lsilvo';
  var opts = {
    url: urlBase + username + '/png'
  };
  request(opts, function (err, response, body) {
    if (err) throw err;
    gt.equal(response.statusCode, 500, 'got 500 response');
    gt.start();
  });
});


src/test/server.js
var listen = require('../app');
var request = require('request');

var port = 4444;
var urlBase = 'http://localhost:' + port + '/';

gt.module('server', {
  setupOnce: function () {
    console.log('starting listening');
    listen(port);
  }
});

gt.async('no username', function () {
  var url = urlBase;
  request(url, function (err, response, body) {
    if (err) throw err;
    gt.equal(response.statusCode, 401, '401 without username');
    gt.start();
  });
});

gt.async('jashkenas', function () {
  var url = urlBase + 'jashkenas';
  request(url, function (err, response, body) {
    if (err) throw err;
    gt.equal(response.statusCode, 200, 'got response');
    gt.ok(/jashkenas/.test(body), 'contains username');
    gt.ok(/Total/i.test(body), 'contains total');
    gt.start();
  });
});



index.js
var connect = require('connect');
var http = require('http');
var proud = require('proud');
var badge = require('proud-badge');
var report = require('./src/report');
var fs = require('fs');
var check = require('check-types');
var Q = require('q');
var moment = require('moment');

var port = process.env.PORT || 3000;

// username -> {image, date}
var badges = {};

function ensureSpace() {
  var maxN = 1000;
  if (Object.keys(badges).length > maxN) {
    console.log('freeing space');
    // todo: sort by date, remove oldest
  }
}

function generateBadge(username) {
  check.verify.unemptyString(username, 'expected username');

  if (badge[username] && badge[username].generating) {
    return badge[username].generating;
  }

  ensureSpace();

  if (!badge[username]) {
    badge[username] = {};
  }

  // just keep promise in the image's placeholder
  badge[username].generating = badge(username)
  .then(function (filename) {
    check.verify.unemptyString(filename, 'expected generated badge filename');
    var image = fs.readFileSync(filename);
    badges[username] = {
      image: image,
      date: new moment()
    };

    fs.unlinkSync(filename);

    return image;
  });

  return badge[username].generating;
}

function getBadge(username) {
  check.verify.unemptyString(username, 'expected username');

  var oldDate = moment().subtract('days', 1);
  if (badges[username] && badges[username].image) {
    if (badges[username].date.isBefore(oldDate)) {
      console.log('badge for', username, 'is out of date');
      return generateBadge(username);
    }

    if (badges[username].image) {
      return badges[username].image;
    }
  }

  return generateBadge(username);
}

function sendBadge(username, res) {
  console.log('returning badge image for', username);
  check.verify.unemptyString(username, 'expected username');

  Q.when(getBadge(username))
  .then(function (image) {
    if (!image) {
      throw new Error('Undefined image for user ' + username);
    }

    res.writeHead(200, {
      'Content-Length': image.length,
      'Content-Type': 'image/png'
    });
    res.write(image);
    res.end();
  })
  .catch(function (err) {
    console.error('Error generating badge for', username);
    console.error(err);
    console.error(err.stack);
    res.writeHead(500, err);
    res.end();
  });
}

function sendTextReport(username, res) {
  console.log('returning text for', username);

  proud(username)
  .then(report.bind(null, username))
  .then(function (report) {
    console.log('report', report);

    if (!report) {
      res.end(username + ' has no modules\n');
    } else {
      res.end(report);
    }
  })
  .catch(console.error);
}

var app = connect()
  .use(connect.favicon())
  .use(connect.logger('dev'))
  .use(connect.static('public'))
  .use(connect.query())
  .use(function (req, res) {
    if (!req.url || req.url === '/') {
      res.writeHead(401, 'missing NPM username');
      res.end();
      return;
    }

    var username = req.url.split('/')[1];
    if (!username) {
      res.writeHead(401, 'missing NPM username');
      res.end();
      return;
    }

    if (/image/.test(req.headers.accept)) {
      sendBadge(username, res);
    } else {
      sendTextReport(username, res);
    }
  });

http.createServer(app).listen(port);
console.log('listening to port', port);


src/app.js
var connect = require('connect');
var http = require('http');
var proud = require('proud');
var badge = require('proud-badge');
var report = require('./report');
var fs = require('fs');
var check = require('check-types');
var Q = require('q');
var moment = require('moment');

// username -> {image, date}
var badges = {};

function ensureSpace() {
  var maxN = 1000;
  if (Object.keys(badges).length > maxN) {
    console.log('freeing space');
    // todo: sort by date, remove oldest
  }
}

function keepGeneratedBadge(username, filename) {
  check.verify.unemptyString(username, 'expected username');
  check.verify.unemptyString(filename, 'expected generated badge filename');

  var image = fs.readFileSync(filename);
  badges[username] = {
    image: image,
    date: new moment()
  };

  fs.unlinkSync(filename);

  return image;
}

function generateBadge(username) {
  check.verify.unemptyString(username, 'expected username');

  if (badge[username] && badge[username].generating) {
    return badge[username].generating;
  }

  ensureSpace();

  if (!badge[username]) {
    badge[username] = {};
  }

  badge[username].generating = badge(username)
    .then(keepGeneratedBadge.bind(null, username));

  return badge[username].generating;
}

function getBadge(username) {
  check.verify.unemptyString(username, 'expected username');

  var oldDate = moment().subtract('days', 1);
  if (badges[username] && badges[username].image) {
    if (badges[username].date.isBefore(oldDate)) {
      console.log('badge for', username, 'is out of date');
      return generateBadge(username);
    }

    if (badges[username].image) {
      return badges[username].image;
    }
  }

  return generateBadge(username);
}

function sendBadge(username, res) {
  console.log('returning badge image for', username);
  check.verify.unemptyString(username, 'expected username');

  Q.when(getBadge(username))
  .then(function (image) {
    if (!image) {
      throw new Error('Undefined image for user ' + username);
    }

    res.writeHead(200, {
      'Content-Length': image.length,
      'Content-Type': 'image/png'
    });
    res.write(image);
    res.end();
  })
  .catch(function (err) {
    console.error('Error generating badge for', username);
    console.error(err);
    console.error(err.stack);
    res.writeHead(500, err);
    res.end();
  });
}

function sendTextReport(username, res) {
  console.log('returning text for', username);

  proud(username)
  .then(report.bind(null, username))
  .then(function (report) {
    res.writeHead(200, {
      'Content-Type': 'text/plain'
    });

    if (!report) {
      res.end(username + ' has no modules\n');
    } else {
      res.end(report);
    }
  })
  .catch(console.error);
}

function sendJsonReport(username, res) {
  console.log('returning json for', username);

  proud(username)
  .then(report.bind(null, username))
  .then(function (report) {
    var result = {
      username: username,
      report: report
    };
    console.log('report', result);

    res.writeHead(200, {
      'Content-Type': 'application/json'
    });

    res.end(JSON.stringify(result));
  })
  .catch(console.error);
}

function isImage(format) {
  return (/image/).test(format) ||
    format === 'png' ||
    format === 'image';
}

var app = connect()
  .use(connect.favicon())
  .use(connect.logger('dev'))
  .use(connect.static('public'))
  .use(connect.query())
  .use(function (req, res) {
    if (!req.url || req.url === '/') {
      res.writeHead(401, 'missing NPM username');
      res.end();
      return;
    }

    var parts = req.url.split('/');
    var username = parts[1];
    if (!username) {
      res.writeHead(401, 'missing NPM username');
      res.end();
      return;
    }

    var acceptsFormat = req.headers.accept;
    if (parts.length > 2) {
      acceptsFormat = parts[2];
      check.verify.unemptyString(acceptsFormat, 'wrong format ' + acceptsFormat);
      console.log('requesting format', acceptsFormat);
    }
    console.log('accepts', acceptsFormat);

    if (isImage(acceptsFormat)) {
      sendBadge(username, res);
    } else if (acceptsFormat === 'application/json' ||
      acceptsFormat === 'json') {
      sendJsonReport(username, res);
    } else {
      sendTextReport(username, res);
    }
  });

module.exports = function listen(port) {
  port = port || process.env.PORT || 3000;
  http.createServer(app).listen(port);
  console.log('listening to port', port);
};
